qrexec-lib/unpack.c

```diff
@@ -227,17 +227,17 @@ void process_one_file(struct file_header *untrusted_hdr)
     untrusted_namebuf[namelen] = 0;
     if (S_ISREG(untrusted_hdr->mode))
         process_one_file_reg(untrusted_hdr, untrusted_namebuf);
-    else if (S_ISLNK(untrusted_hdr->mode))
+    else if (S_ISLNK(untrusted_hdr->mode) && (flags & COPY_ALLOW_SYMLINKS))
         process_one_file_link(untrusted_hdr, untrusted_namebuf);
-    else if (S_ISDIR(untrusted_hdr->mode))
+    else if (S_ISDIR(untrusted_hdr->mode) && (flags & COPY_ALLOW_DIRECTORIES))
         process_one_file_dir(untrusted_hdr, untrusted_namebuf);
     else
         do_exit(EINVAL, untrusted_namebuf);
     if (verbose && !S_ISDIR(untrusted_hdr->mode))
         fprintf(stderr, "%s\n", untrusted_namebuf);
 }
 
-int do_unpack(void)
+int do_unpack(int flags)
```

Comment:
This API change depends on what we want to do with the default case (`qubes.FileCopy` service). 
I think the preferable option is to keep the `qubes.FileCopy` service as is (allowing dirs and symlinks there), and optionally introduce another service `qubes.FileCopySimple` (or maybe an argument to `qubes.FileCopy`?) that limits what can be copied. If user choose to use this simpler version, the sending side can filter what is sent, possibly also listing things that were skipped (or abort with meaningful message). When it's rejected at the receiving side, the error message will be much less meaningful.

If going this way, I'd propose to rename `do_unpack(void)` to `do_unpack_flags(int flags)` (or `_ext`?), then add `do_unpack(void)` as a wrapper for the `do_unpack_flags(COPY_ALLOW_DIRECTORIES | COPY_ALLOW_SYMLINKS)`. This way we avoid the need for soname bump, which simplifies the upgrade (no need to rebuild things linked with it).

Another option (IMO less preferable) is to change `qubes.FileCopy` to reject symlinks by default (but allow directories). In this case indeed flags can be added to do_unpack() directly, as the sender will need to be adjusted for this too. But it will result in confusing errors, if one uses older sender side.

Reply:
What if we:

- add an argument to `qubes.Filecopy` that specifies the allowed types to unpack
- have no argument be the current behavior
- deprecate and eventually remove the legacy behavior?

Reply:
For the underlying `qfile-unpacker` tool that could work indeed. But I don't think exposing this choice in full to the user make much sense. It isn't realistic for the user to choose _for each copy operation_ what type of objects are there (out of all 2x2 options). It could be automated on the source side based on actual object types, but then limiting it doesn't make much sense, if the source can choose what to request (many users won't notice such difference). Furthermore, forbidding directories for normal `qubes.FileCopy` is something that is almost never the right thing to do.

So, as for different usages of this code goes:
 - `qubes.FileCopy` - should have a variant that forbids symlinks (maybe an argument?), but have an option to allow symlinks too
 - `qubes.ReceiveUpdates` - should forbid both directories and symlinks
 - retrieving backup (the other side of `qubes.Restore`) - should forbid symlinks